### PR TITLE
[KYUUBI #1643] Implement GetFunctions operation

### PR DESCRIPTION
**Pull request by SteNicholas**
### _Why are the changes needed?_

Implement GetFunctions operation.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
Thanks for contribution! Could you please rebase you branch on latest master? BTW, please run `dev/reformat` before commit your code, you can also configure your IDEA use scalafmt to formatter your code rather than the default one.

**Comment by SteNicholas** [comment]
@pan3793 , sorry for not rebasing the master branch. I have rebased the latest master branch. Please help to review this pull request.

**Comment by SteNicholas** [comment]
@pan3793 , sorry for not using the `dev/reformat`. I have rebased the latest master branch and reformated the codes. Please help to review this pull request.

**Comment by yaooqinn** [comment]
cc @yanghua @link3280

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      var rows: List[Row] = List()
+      val tableEnv: TableEnvironment = sessionContext.getExecutionContext.getTableEnvironment
+      tableEnv.listFunctions().diff(tableEnv.listUserDefinedFunctions())
+        .filter(f => StringUtils.isEmpty(functionName) || f == functionName)
+        .foreach { f =>
+          columns += (FUNCTION_NAME -> Tuple2(math.max(f.length, columns(FUNCTION_NAME)._1), true))
+          rows = Row.of(null, null, f) +: rows
+        }
+      tableEnv.listCatalogs()
+        .filter(c => StringUtils.isEmpty(catalogName) || c == catalogName)
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:55]: I think it should be pattern match not exactly string match

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        .foreach { c =>
+          columns += (FUNCTION_CAT -> Tuple2(math.max(c.length, columns(FUNCTION_CAT)._1), true))
+          val catalog = tableEnv.getCatalog(c).get()
+          catalog.listDatabases().asScala
+            .filter(d => StringUtils.isEmpty(schemaName) || d == schemaName)
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:60]: ditto

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            .foreach { d =>
+              columns += (
+                FUNCTION_CAT -> Tuple2(math.max(d.length, columns(FUNCTION_SCHEM)._1), true))
+              catalog.listFunctions(d).asScala
+                .filter(f => StringUtils.isEmpty(functionName) || f == functionName)
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:65]: ditto

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        }
+      resultSet = ResultSet.builder.resultKind(ResultKind.SUCCESS_WITH_CONTENT)
+        .columns(
+          Column.physical(
+            columns.head._1,
+            if (columns.head._2._2) DataTypes.VARCHAR(columns.head._2._1).notNull()
+            else DataTypes.VARCHAR(columns.head._2._1)),
+          Column.physical(
+            columns.toList(1)._1,
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:80]: Do we create multiple lists by calling toList multiple times?

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    schemaName: String,
+    functionName: String)
+  extends FlinkOperation(OperationType.GET_FUNCTIONS, session) {
+
+  override protected def runInternal(): Unit = {
+    try {
+      var columns: Map[String, (Int, Boolean)] = Map(
+        FUNCTION_CAT -> Tuple2(VarCharType.DEFAULT_LENGTH, false),
+        FUNCTION_SCHEM -> Tuple2(VarCharType.DEFAULT_LENGTH, false),
+        FUNCTION_NAME -> Tuple2(VarCharType.DEFAULT_LENGTH, false))
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:45]: I think function names must not be empty or null in any case. WDYT?

**Review by link3280 — COMMENTED** [review]
@SteNicholas Thanks a lot for your work! I left some comments, please kindly take a look. 

Besides that, I'm not very sure if we should determine column types (e.g. its scale or precision) based on the returned data. I think types should reflect all possible values of the query, but now we may get `FUNTION_NAME varchar(16)` at one result set, `FUNTION_NAME varchar(20)` from another. WDYT?

**Comment by SteNicholas** [comment]
> @SteNicholas Thanks a lot for your work! I left some comments, please kindly take a look.
> 
> Besides that, I'm not very sure if we should determine column types (e.g. its scale or precision) based on the returned data. I think types should reflect all possible values of the query, but now we may get `FUNTION_NAME varchar(16)` at one result set, `FUNTION_NAME varchar(20)` from another. WDYT?

@link3280 , I refer to the implementation of the `GetCatalogs` operation worked by @yanghua . In `GetCatalogs` operation, the column types are determined by the returned data. IMO, the colum types determined by data makes sense to me because the data types are actually get by the actual data.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala (hunk 5 above)
Comment (SteNicholas) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:45]: Yes, I agree with the point that the `FUNCTION_NAME` column couldn't be nullable. I have set this non nullable after getting the functions.

**Comment by yaooqinn** [comment]
I am +1 with @link3280, the get catalog API shall have a static result schema including column size, name, type, and restraint. Otherwise, we will result in supersize for clients like HUE/tableu/dbeaver

**Comment by link3280** [comment]
@SteNicholas +1 to `STRING`.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      val schemaPattern = toJavaRegex(schemaName)
+      val functionPattern = toJavaRegex(functionName)
+      var rows: List[Row] = List()
+      val tableEnv: TableEnvironment = sessionContext.getExecutionContext.getTableEnvironment
+      filterPattern(
+        tableEnv.listFunctions().diff(tableEnv.listUserDefinedFunctions()),
+        functionPattern)
+        .foreach { f =>
+          rows = Row.of(null, null, f) +: rows
+        }
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:51]: We can leverage `map` `flatMap` to simplify the code.
```suggestion
    val globalFunctions = filterPattern(
        tableEnv.listFunctions().diff(tableEnv.listUserDefinedFunctions()),
        functionPattern)
        .map { f => Row.of(null, null, f) }
```

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      tableEnv.listCatalogs()
+        .filter { c => StringUtils.isEmpty(catalogName) || c == catalogName }
+        .foreach { c =>
+          val catalog = tableEnv.getCatalog(c).get()
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:55]: Your filter accepts `null`, will it cause NPE if catalogs contains `null`?

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.flink.util
+
+import java.util.regex.PatternSyntaxException
+
+import org.apache.kyuubi.Logging
+
+object StringUtils extends Logging {
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala:24]: Seems `extends Logging` is not required

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * This utility can be used for filtering pattern in the "Like" of "Show Tables / Functions" DDL
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala:27]: Let's add comments where this function come from, and if possible, also add a ut to help user understand how it works.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+   * @param names the names list to be filtered
+   * @param pattern the filter pattern, only '*' and '|' are allowed as wildcards, others will
+   *                follow regular expression convention, case insensitive match and white spaces
+   *                on both ends will be ignored
+   * @return the filtered names list in order
+   */
+  def filterPattern(names: Seq[String], pattern: String): Seq[String] = {
+    val funcNames = scala.collection.mutable.SortedSet.empty[String]
+    pattern.trim().split("\\|").foreach { subPattern =>
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/util/StringUtils.scala:36]: What about regex patterns that contain '|' or '*'? It seems that they're not properly escaped.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  extends FlinkOperation(OperationType.GET_FUNCTIONS, session) {
+
+  override protected def runInternal(): Unit = {
+    try {
+      val schemaPattern = toJavaRegex(schemaName)
+      val functionPattern = toJavaRegex(functionName)
+      var rows: List[Row] = List()
+      val tableEnv: TableEnvironment = sessionContext.getExecutionContext.getTableEnvironment
+      filterPattern(
+        tableEnv.listFunctions().diff(tableEnv.listUserDefinedFunctions()),
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/GetFunctions.scala:47]: Do we have a consensus that we need built-in system functions only? Docs of [DatabaseMetaData](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html#getFunctions-java.lang.String-java.lang.String-java.lang.String-) that you refer to states user functions should be included as well.

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala [hunk 12]
```diff
@@ -119,6 +122,36 @@ class FlinkOperationSuite extends WithFlinkSQLEngine with HiveJDBCTestHelper {
...
+              assert(resultSet.getString(FUNCTION_CAT) == null)
+              assert(resultSet.getString(FUNCTION_SCHEM) === null)
+              assert(resultSet.getString(FUNCTION_NAME) === func)
+            }
+          }
+        }
+      val expected =
+        List("currentTimestamp", "currentDate", "currentTime", "localTimestamp", "localTime")
+      Seq("current", "local")
+        .foreach { func =>
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:142]: `func` -> `pattern`?

**Review by link3280 — COMMENTED** [review]
@SteNicholas Thanks for your contribution. I left some minor comments, PTAL.

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**